Use std::chrono for the time to rotate destination of addr messages + tests

naumenkogs opened the pull request:
We currently assign a destination peer for relaying particular addresses of nodes every 24 hours, and then rotate. This is done for rate-limiting (ultimately for privacy leak reduction I think?).

Before this change, 24 hours was defined as uint. I replaced it with std::chrono, which is mockable and type-safe.

Also added couple tests for this behavior.

naumenkogs committed:
Use std::chrono for salting when randomizing ADDR destination

naumenkogs committed:
Add tests for addr destination rotation

Check that within 24h addr of a given node is forwarded
to the same peer(s), and then the destination is
rotated every 24h.

Co-authored-by: Jon Atack <jon@atack.com>

sipa commented on src/net_processing.cpp:
```diff
@@ -1373,7 +1375,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
     uint64_t hashAddr = addr.GetHash();
-    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr << 32).Write((GetTime() + hashAddr) / (24 * 60 * 60));
+
+    std::chrono::seconds rotate_interval = std::chrono::duration_cast<std::chrono::seconds>(ROTATE_ADDR_RELAY_DEST_INTERVAL);
```
Is this duration_cast needed? I think conversion should be automatic.

naumenkogs replied:
You are right. I spent 30 minutes fighting with this code, and couldn't come up with something better.

But after you messaged I tried once more and it worked. Will force push in a second.

naumenkogs commented:
@sipa btw, take a look at how hashAddr was fed to the hasher twice (with a shift, yeah).
I removed it, although I don't know why it was there in the first place... Feeding it once like I do should be sufficient.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1373,7 +1375,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
+
+    std::chrono::seconds rotate_interval = std::chrono::duration_cast<std::chrono::seconds>(ROTATE_ADDR_RELAY_DEST_INTERVAL);
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    uint64_t time_seed = current_time.count() / rotate_interval.count();
```
```suggestion
    const uint64_t time_seed {GetTime<std::chrono::seconds>().count() / std::chrono::seconds{ROTATE_ADDR_RELAY_DEST_INTERVAL}.count()};
```

Everything in one line with compile-time checks enabled would look like this, btw

naumenkogs replied:
can we keep it to 2 lines? I think what you suggest is hard to read.

sipa commented:
@naumenkogs It was very inscrutable code, but the construction with double-hashing the address hash was by design.

The idea is that the address both influences the hash directly, but also determines when the "reset time of the day" is (by including hashaddr in `(time + hashaddr)/DAY`, not every address will change salt simultaneously throughout the day).

The shifting is pointless, we should get rid of it. It seems to be a silly evolution of this 2010 Satoshi code: 5cbf75324d1509a1262b65c5073314a4da3f6d77.

sipa commented on src/net_processing.cpp:
```diff
@@ -1373,7 +1375,10 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    uint64_t time_seed = current_time / ROTATE_ADDR_RELAY_DEST_INTERVAL;
+    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr).Write(time_seed);
```
```suggestion
    const CSipHasher hasher = connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr).Write((current_time + hashAddr) / ROTATE_ADDR_RELAY_DEST_INTERVAL);
```

naumenkogs replied:
Yeah, I'm not 100% sure that this measure is effective against anything, but it won't hurt to have a less symmetry.
This mehaviour _may_ leak extra information if someone tries to analyze how addresses are flowing, but since we don't have any real evidence, we probably should stick to the existing reasoning.

Had to slightly adjust the code you suggested, because adding uint to chrono type doesn't work right away.

naumenkogs commented:
Updated the code with suggestion by @sipa, and simplified tests.

They are a bit loosened now, but more robust to failures caused by randomness. 
Loosened in a sense they don't check the exact *daily* logic, because we can't predict when rotation happens for a given address (well, we can, but it's probably not worth to re-code all choose destination logic in python)

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1373,7 +1375,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = (current_time + std::chrono::hours(hashAddr)) / ROTATE_ADDR_RELAY_DEST_INTERVAL;
```
Previously `hashAddr` was of type seconds, now it is type hours. What was the reason to change it?

naumenkogs replied:
I think i changed it while trying something else. There is no real difference, as long as it results is on order of hours or more. Otherwise, it won't add enough randomness.

MarcoFalke replied:
You will get rid of the "reset time of the day" feature. Generally I don't like refactoring commits to also change behavior.

naumenkogs replied:
Okay, made it seconds again so that we don't change the behavior. Although I think we won't loose any properties if hours are used.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1373,7 +1375,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = (current_time + std::chrono::seconds(hashAddr)) / ROTATE_ADDR_RELAY_DEST_INTERVAL;
```
```suggestion
    uint64_t time_addr = (current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL;
```

Maybe try this to avoid https://travis-ci.org/github/bitcoin/bitcoin/jobs/675780723#L4756

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1373,7 +1375,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = (current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL;
```
```suggestion
    uint64_t time_addr = static_cast<uint64_t>((current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL);
```


```
 node0 stderr net_processing.cpp:1381:26: runtime error: implicit conversion from type 'typename common_type<long, long>::type' (aka 'long') of value -28739051454386 (64-bit, signed) to type 'uint64_t' (aka 'unsigned long') changed the value to 18446715334658097230 (64-bit, unsigned)

jonatack commented on test/functional/p2p_addr_relay.py:
```diff
@@ -48,24 +53,72 @@ def run_test(self):
             addr_receiver.sync_with_ping()
 
+        self.log.info('Check that within 24 hours an addr relay destination is rotated at most once')
+        mocked_time = int(time.time()) + 100000
```
I don't grok what/why regarding the 100000 value. It would be nice to comment and/or hoist it to a variable.

Same for 200000 at line 100.

jonatack commented on test/functional/p2p_addr_relay.py:
```diff
@@ -48,24 +53,72 @@ def run_test(self):
+                for receiver in addr_receivers:
+                    receiver.sync_with_ping()
+        nodes_received_addr = [node for node in addr_receivers if node.received_addr]
+        assert(len(nodes_received_addr) <= 4)
```
Why the value of 4 here and line 118? Perhaps add a comment.

jonatack reviewed (commented):
(Tested) Concept ACK.

Some questions and style/PEP8 comments below, feel free to ignore.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -1433,7 +1435,11 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = static_cast<uint64_t>((current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL);
```
We iterated over this line a couple times with @MarcoFalke, so I'm out of ideas here.
Open to suggestions!

sipa replied:
Annoyingly, I think this is UB, as hashAddr may exceed 2^63, and casting a larger number to a signed 64-bit integer would be outside its range (and overflowing signed integers is undefined behavior).

Suggestion: `uint64_t time_addr = (current_time + std::chrono::seconds{hashAddr % ROTATE_ADDR_RELAY_DEST_INTERVAL.count()}) / ROTATE_ADDR_RELAY_DEST_INTERVAL;`, and make `ROTATE_ADDR_RELAY_DEST_INTERVAL` to be of type `std::chrono::seconds`.

sipa replied:
This isn't resolved actually; it's still casting `hashAddr` from `uint64_t` to `int64_t`, which is implementation-defined behavior (I was wrong claiming earlier that it was UB).

sdaftuar replied:
We're constructing a `std::chrono::seconds` from an `int64_t`, but according to https://en.cppreference.com/w/cpp/chrono/duration, `std::chrono::seconds` is only guaranteed to have 35 bits -- is this problematic?

sipa replied:
Uh, right. I think that could be a problem. This will need something like

```c++
uint64_t time_addr = (uint64_t{current_time / std::chrono::seconds{1}} + hashAddr) / (ROTATE_ADDR_RELAY_DEST_INTERVAL / std::chrono::seconds{1});
```

(I prefer using `x / std::chrono::seconds{1}` over `x.count()` to get the number of seconds of `x`, as the former expression remains meaningful if the type of `x` changes to something other than `std::chrono::seconds`).

MarcoFalke replied:
My understanding was that `std::chrono::seconds{int64_t{}}` wouldn't compile on platforms where seconds isn't backed by at least 64 bits. I don't expect any such platform to exist.

@ sipa there is also a `count_seconds()` helper.

naumenkogs replied:
Github is weird, I found this un-resolved issue totally accidentally...

naumenkogs replied:
Okay the latest version should be safe?

naumenkogs commented:
@jonatack addressed all the things.

jonatack reviewed (commented):
ACK 72d7e24

Thanks for clarifying the tests! I spent some time messing around with them, refactored, and sped them up from 26 to 7 seconds on my machine. Feel free to use/steal/ignore: https://github.com/jonatack/bitcoin/commits/refactor-p2p_addr_relay-tests

sipa commented:
utACK 72d7e2400759dacfc8860341d5bb90459e363a81. I only casually reviewed the tests.

MarcoFalke commented:
ACK 72d7e2400759dacfc8860341d5bb90459e363a81

fanquake commented:
Aside from this needing rebase for conflicts, can you rebase on master so that new CIs run. That should remove appveyor as well.

MarcoFalke commented:
@naumenkogs Is this up for grabs?

naumenkogs commented:
Okay, it was hard to rebase @jonatack test contributed, but I just did it. Ready for review.

theStack reviewed (commented):
Concept ACK

The first commits replaces `GetTime()` and the timing constants related to the randomization of ADDR messages with the mockable variants, i.e. it enables mocking but there shouldn't be any change in behaviour if mocking is not used. Will take a deeper look on the test soon, just left some nits (feel free to ignore).

MarcoFalke commented:
@naumenkogs Is this up for grabs?

naumenkogs commented:
@MarcoFalke I mean, it's up-to-date (apart from those little nits) and reviewable.
I probably don't have energy to nudge people for reviewing this, so if that makes it "up for grabs", then yes.

MarcoFalke commented:
It would be good to reply to the comments. Either addressing them or denying/closing them.

MarcoFalke commented:
Or alternatively submit the test in a separate pull request, since this is refactoring, so the test should already pass on master.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -1700,7 +1702,11 @@ void PeerManagerImpl::RelayAddress(NodeId originator,
+
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = static_cast<uint64_t>((current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL);
```
Can you explain why it is safe / defined behavior to perform a `static_cast<uint64_t>(std::chrono::seconds)`?  I can't figure out from reading https://en.cppreference.com/w/cpp/chrono/duration whether this is supposed to be defined, but maybe I'm missing something.

sdaftuar replied:
Sorry for the noise, @sipa explained this to me (the result of the division is being cast, which is not a duration but a number).

sipa commented on src/net_processing.cpp:
```diff
@@ -1700,7 +1702,11 @@ void PeerManagerImpl::RelayAddress(NodeId originator,
+    std::chrono::seconds current_time = GetTime<std::chrono::seconds>();
+    // Adding address hash makes exact rotation time different per address, while preserving periodicity.
+    uint64_t time_addr = static_cast<uint64_t>((current_time + std::chrono::seconds{static_cast<int64_t>(hashAddr)}) / ROTATE_ADDR_RELAY_DEST_INTERVAL);
+    const CSipHasher hasher = m_connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr << 32).Write(time_addr);
```
No need for the `<< 32` here, I think.

MarcoFalke replied:
I've pulled that out into #23970, since it also resolves a sanitizer issue.

naumenkogs replied:
@MarcoFalke does it mean i can leave it as it is here?

MarcoFalke replied:
You can either leave it as is or take it, but if you take it, make sure to also remove the suppression like in #23970.

MarcoFalke commented:
Just to note `GetTime` is already mockable, so this only changes the type. Though, if this is too much hassle to get right and too hard to review (due to static_cast and `{}`-cast hell), then maybe leave the code as is and only add the test?

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -124,7 +124,9 @@ static const unsigned int NODE_NETWORK_LIMITED_MIN_BLOCKS = 288;
 static constexpr auto AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL = 24h;
 /** Average delay between peer address broadcasts */
-static constexpr auto AVG_ADDRESS_BROADCAST_INTERVAL = 30s;
+static constexpr std::chrono::seconds AVG_ADDRESS_BROADCAST_INTERVAL{30};
```
why is this changed? This also conflicts with #23832

naumenkogs replied:
I think it was like that from the very beginning? But sure, I can drop this.

sipa replied:
Using the `static constexpr auto NAME = <numer><suffix>;` as is used here is a bit more readable I think. I'd suggest using the same for the line below.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1746,8 +1748,12 @@ void PeerManagerImpl::RelayAddress(NodeId originator,
     // at a time so the m_addr_knowns of the chosen nodes prevent repeats
-    const uint64_t hashAddr{addr.GetHash()};
-    const CSipHasher hasher{m_connman.GetDeterministicRandomizer(RANDOMIZER_ID_ADDRESS_RELAY).Write(hashAddr).Write((GetTime() + hashAddr) / (24 * 60 * 60))};
+    uint64_t hashAddr = addr.GetHash();
```
why is this changed?

jonatack replied:
left that line unchanged in the rebase at https://github.com/jonatack/bitcoin/commit/77ccb7fce15e340080f14c7626cf3dc63fcdee88

jonatack reviewed (commented):
@naumenkogs if helpful, here is a rebase of this pull: https://github.com/jonatack/bitcoin/commits/pr18642-rebase-to-master-at-e04720e. If you don't have time to update this in a week or so, I'll open a PR for you with the same commits.

jonatack reviewed (commented):
ACK 2ff8f4dd81dc484fe38ddd9db63cc8fd30192245
